File: leetcode/leetcode_cpp/flatten-binary-tree-to-linked-list.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

 /*
 https://leetcode.com/problems/flatten-binary-tree-to-linked-list

 1. dfs (post order)
 2. no recursion and stack(queue)
 
 */
class Solution {
public:
    void dfs(TreeNode* node) {
        if (!node) return;
        
        // left
        dfs(node->left);
        dfs(node->right);
        
        if (node->left) {
            auto cur = node->left;
            while (cur->right) cur = cur->right;
            cur->right = node->right;
            node->right = node->left;
            node->left = nullptr;
        }
    }
    
    void flatten_1(TreeNode* root) {
        dfs(root);
    }

    void flatten_2(TreeNode* root) {
        while (root) {
            if (root->left && root->right) {
                auto t = root->left;
                while (t->right)
                    t = t->right;
                t->right = root->right;
            }

            if (root->left) {
                root->right = root->left;
            }
            root->left = nullptr;
            root = root->right;
        }
    }

    void flatten(TreeNode* root) {
        flatten_2(root);
    }
};

File: leetcode/leetcode_py/flatten-binary-tree-to-linked-list.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
'''
https://leetcode.com/problems/flatten-binary-tree-to-linked-list

1. dfs (post order)
2. no recursion and stack(queue)
 
'''
class Solution:
    def flatten_1(self, root: Optional[TreeNode]) -> None:
        if not root: return
        self.flatten_1(root.left)
        self.flatten_1(root.right)
        if root.left:
            cur = root.left
            while cur.right: cur = cur.right
            cur.right = root.right
            root.right = root.left
            root.left = None

    def flatten_2(self, root: Optional[TreeNode]) -> None:
        if not root: return
        while root:
            if root.left and root.right:
                cur = root.left
                while cur.right: cur = cur.right
                cur.right = root.right
                
            if root.left:
                root.right = root.left
            root.left = None
            root = root.right
        
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        self.flatten_2(root) 
        